core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -140,157 +140,164 @@ public void startRecovery(final IndexShard indexShard, final DiscoveryNode sourc
-        logger.trace("will retry recovery with id [{}] in [{}] (reason [{}])", recoveryRef.status().recoveryId(), retryAfter, reason);
-        retryRecovery(recoveryRef, retryAfter, retryCleanupTimeout, currentRequest);
+    protected void retryRecovery(final long recoveryId, final String reason, TimeValue retryAfter, TimeValue activityTimeout) {
+        logger.trace("will retry recovery with id [{}] in [{}] (reason [{}])", recoveryId, retryAfter, reason);
+        retryRecovery(recoveryId, retryAfter, activityTimeout);
     }
 
-    private void retryRecovery(final RecoveryRef recoveryRef, TimeValue retryAfter, TimeValue retryCleanupTimeout,
-                               final StartRecoveryRequest currentRequest) {
-        RecoveryTarget newTarget = onGoingRecoveries.resetRecovery(recoveryRef, retryCleanupTimeout, currentRequest);
+    private void retryRecovery(final long recoveryId, TimeValue retryAfter, TimeValue activityTimeout) {
+        RecoveryTarget newTarget = onGoingRecoveries.resetRecovery(recoveryId, activityTimeout);
         if (newTarget != null) {
             threadPool.schedule(retryAfter, ThreadPool.Names.GENERIC, new RecoveryRunner(newTarget.recoveryId()));
         }
     }
 
-    private void doRecovery(final RecoveryRef recoveryRef) {
-        RecoveryTarget recoveryTarget = recoveryRef.status();
-        assert recoveryTarget.sourceNode() != null : "can't do a recovery without a source node";
-
-        logger.trace("collecting local files for {}", recoveryTarget);
-        Store.MetadataSnapshot metadataSnapshot = null;
-        try {
-            if (recoveryTarget.indexShard().indexSettings().isOnSharedFilesystem()) {
-                // we are not going to copy any files, so don't bother listing files, potentially running
-                // into concurrency issues with the primary changing files underneath us.
-                metadataSnapshot = Store.MetadataSnapshot.EMPTY;
-            } else {
-                metadataSnapshot = recoveryTarget.indexShard().snapshotStoreMetadata();
-            }
-        } catch (org.apache.lucene.index.IndexNotFoundException e) {
-            // happens on an empty folder. no need to log
-            metadataSnapshot = Store.MetadataSnapshot.EMPTY;
-        } catch (IOException e) {
-            logger.warn("error while listing local files, recover as if there are none", e);
-            metadataSnapshot = Store.MetadataSnapshot.EMPTY;
-        } catch (Exception e) {
-            // this will be logged as warning later on...
-            logger.trace("unexpected error while listing local files, failing recovery", e);
-            onGoingRecoveries.failRecovery(recoveryTarget.recoveryId(),
-                    new RecoveryFailedException(recoveryTarget.state(), "failed to list local files", e), true);
+    private void doRecovery(final long recoveryId) {
+        RecoveryRef recoveryRef = onGoingRecoveries.getRecovery(recoveryId);
```

Comment:
we should release the reference. Can we do all prep work under a try with resources until and including the creation of the start recovery requests? I believe we only need the recovery id and the cancelable threads after that, so we can capture it within the block. It will be easier to follow.